src/Notification/Job/SendNotificationsJob.php

```diff
@@ -0,0 +1,66 @@
+<?php
+
+namespace Flarum\Notification\Job;
+
+use Carbon\Carbon;
+use Flarum\Notification\Blueprint\BlueprintInterface;
+use Flarum\Notification\Event\Notifying;
+use Flarum\Notification\Event\Sending;
+use Flarum\Notification\MailableInterface;
+use Flarum\Notification\Notification;
+use Flarum\Notification\NotificationMailer;
+use Flarum\Notification\NotificationSyncer;
+use Flarum\Queue\AbstractJob;
+use Flarum\User\User;
+
+class SendNotificationsJob extends AbstractJob
+{
+    /**
+     * @var BlueprintInterface
+     */
+    private $blueprint;
+    /**
+     * @var array
+     */
+    private $recipientIds;
+
+    public function __construct(BlueprintInterface $blueprint, array $recipientIds = [])
+    {
+        $this->blueprint = $blueprint;
+        $this->recipientIds = $recipientIds;
+    }
+
+    public function handle(NotificationMailer $mailer)
+    {
+        $now = Carbon::now('utc')->toDateTimeString();
+        $recipients = $this->recipientIds;
+
+        event(new Sending($this->blueprint, $recipients));
+
+        $attributes = NotificationSyncer::getAttributes($this->blueprint);
+
+        Notification::insert(
+            array_map(function (User $user) use ($attributes, $now) {
+                return $attributes + [
+                        'user_id' => $user->id,
+                        'created_at' => $now
+                    ];
+            }, $recipients)
+        );
+
+        event(new Notifying($this->blueprint, $recipients));
+
+        if ($this->blueprint instanceof MailableInterface) {
+            $this->email($mailer, $this->blueprint, $recipients);
+        }
+    }
+
+    protected function email(NotificationMailer $mailer, MailableInterface $blueprint, array $recipients)
+    {
+        foreach ($recipients as $user) {
+            if ($user->shouldEmail($blueprint::getType())) {
+                $mailer->send($blueprint, $user);
```

Comment:
Do we want to queue these separately? Otherwise, a late failure here would probably cause a lot of re-sends when this job is retried...

Reply:
I considered the same thing. If we want to harden this right now there are two options:

- Dispatch them separately.
- Catch exceptions and ignore them.

..

Reply:
@luceos already changed this so they are now dispatched separately in the `NotificationSyncer`. Works fine for the moment.